src/udp_load_balancer.c

```diff
@@ -21,36 +26,130 @@ unsigned int get_servers_amount(char **servers_list) {
+                       servers_list_resolved_tmp = realloc(servers_list_resolved_tmp, sizeof (char*) * (j+1));
+                       servers_ports_resolved_tmp = realloc(servers_ports_resolved_tmp, sizeof (int) * (j+1));
+                    }
+                    servers_list_resolved_tmp[j] = strdup(inet_ntoa(**addr_list));
+                    servers_ports_resolved_tmp[j] = servers_ports[i];
+                    if ( verb_flag ) printf("    %s %d\n",servers_list_resolved_tmp[j], servers_ports_resolved_tmp[j]);
+                    j++;
+                    addr_list += 1;
+                }
+            } else {
+              if ( j >= servers_amount ) {
+                 servers_list_resolved_tmp = realloc(servers_list_resolved_tmp, sizeof (char*) * (j+1));
+                 servers_ports_resolved_tmp = realloc(servers_ports_resolved_tmp, sizeof (int) * (j+1));
+              }
+              servers_list_resolved_tmp[j] = strdup(servers_hosts[i]);
+              servers_ports_resolved_tmp[j] = servers_ports[i];
+              j++;
+            }
+            servers_amount_resolved = j;
         }
+        if ( j < 1 ) {
+             printf( "Can't resolve any of servers, waiting for 5sec\n" );
+             sleep(5);
+             continue ;
+        }
+        if ( first_run == 0 ) {
+           pthread_mutex_lock(&resolv_mutex);
+        }
+        first_run = 0;
+        servers_list_resolved_free = servers_list_resolved;
+        servers_list_resolved = servers_list_resolved_tmp;
+        servers_ports_resolved_free = servers_ports_resolved;
+        servers_ports_resolved = servers_ports_resolved_tmp;
+        pthread_mutex_unlock(&resolv_mutex);
+        if ( debug_flag ) {
+          printf("servers_list_resolved array:\n");
+          for ( j = 0 ; j < servers_amount_resolved ; j++ ) {
+               printf(" (%u %s:%d)", j, servers_list_resolved_tmp[j], servers_ports_resolved_tmp[j]);
+               printf("\n");
+          }
+        }
+        free_servers_hosts(servers_list_resolved_free,servers_amount_resolved_prev);
+        free(servers_ports_resolved_free);
+    
+        usleep(5000000); // resolve every 5sec
     }
+    free_servers_hosts(servers_hosts,servers_amount);
+    free_servers_ports(servers_ports,servers_amount);
+}
 
-    while(1) {
-        for(i = 0; i < servers_amount; i++) {
-            listen_udp_packet(message, &message_length, socket_);
+void* run_dispatcher( void* call_arg )
+{
+    dispatcher_arg_struct_t *arg = call_arg; 
+    char message[MAX_PAYLOAD_LENGTH];
+    unsigned int message_length;
+    char srv[16];
+    int port;
 
-            send_udp_packet(servers_hosts[i], servers_ports[i], message, message_length, socket_);
+    unsigned int i;
+
+    i = 0;
+    while(1) {
+        if ( debug_flag ) printf("while waiting for a resolv_mutex\n");
+        listen_udp_packet(message, &message_length, arg->socket_);
+        if ( debug_flag ) printf("message:%s ", message );
```

Comment:
shouldn't have a \n?

Reply:
When I was testing this with logger(1), the message had '\n' by itself. But may be it makes sense to check if message already has '\n'- don't add '\n', if not - add '\n'